contrib/envoyproxy/envoy/fakehttp.go

```diff
@@ -0,0 +1,189 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2024 Datadog, Inc.
+
+package envoy
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+
+	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	extproc "github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3"
+	"google.golang.org/grpc/metadata"
+)
+
+// checkPseudoRequestHeaders Verify the required HTTP2 headers are present
+// Some mandatory headers need to be set. It can happen when it wasn't a real HTTP2 request sent by Envoy,
+func checkPseudoRequestHeaders(headers map[string]string) error {
+	for _, header := range []string{":authority", ":scheme", ":path", ":method"} {
+		if _, ok := headers[header]; !ok {
+			return fmt.Errorf("missing required headers: %q", header)
+		}
+	}
+
+	return nil
+}
+
+// checkPseudoResponseHeaders Verify the required HTTP2 headers are present
+// Some mandatory headers need to be set. It can happen when it wasn't a real HTTP2 request sent by Envoy,
+func checkPseudoResponseHeaders(headers map[string]string) error {
+	if _, ok := headers[":status"]; !ok {
+		return fmt.Errorf("missing required ':status' headers")
+	}
+
+	return nil
+}
+
+func getRemoteAddr(md metadata.MD) string {
+	xfwd := md.Get("x-forwarded-for")
+	length := len(xfwd)
+	if length == 0 {
+		return ""
+	}
+
+	// Get the first right value of x-forwarded-for headers
+	// The rightmost IP address is the one that will be used as the remote client IP
+	// https://datadoghq.atlassian.net/wiki/spaces/TS/pages/2766733526/Sensitive+IP+information#Where-does-the-value-of-the-http.client_ip-tag-come-from%3F
+	return xfwd[length-1]
+}
+
+// partitionPeusdoHeaders Separate normal headers of the initial request made by the client and the pseudo headers of HTTP/2
+// - Format the headers to be used by the tracer as a map[string][]string
+// - Set headers keys to be canonical
+func partitionPeusdoHeaders(receivedHeaders []*corev3.HeaderValue) (map[string][]string, map[string]string) {
```

Comment:
```suggestion
func splitPseudoHeaders(receivedHeaders []*corev3.HeaderValue) (headers map[string][]string, pseudoHeaders map[string]string) {
```

(named return parameters here are useful for readability since types don't give much information about what is returned)

Reply:
Applied in db73a1f13bf8855a9c1a351a685f38629d5926bc thanks 😄